dlme_airflow/drivers/json.py

```diff
@@ -1,155 +1,246 @@
-    See https://github.com/h2non/jsonpath-ng#jsonpath-syntax for what constitues a valid JSONPath.
-
-    For example if your Intake catalog contains:
-
-        args:
-            collection_url: https://www.loc.gov/collections/persian-language-rare-materials/?c=100&fo=json
-        metadata:
-          record_selector: "content.results"
-          fields:
-            id:
-              path: "id",
-            title:
-              path: "title"
-            thumbnail:
-              path: "resources[0].image"
-
-    and the collection URL returns JSON that looks like:
-
-    {
-      "content": {
-        "results": [
-          {
-            "id": "http://www.loc.gov/item/2017498321/",
-            "title": "[Majmuʻah, or, Collection]",
-            "resources": [
-              "image": "https://tile.loc.gov/image-services/iiif/service:amed:plmp:m154:0334/full/pct:6.25/0/default.jpg"
-            ]
-          },
-          {
-            "id": "http://www.loc.gov/item/00313408/",
-            "title": "Sakīnat al-fuz̤alāʼ mawsūm bi-ism-i tārīkhī-i Bahār-i Afghānī",
-            "resources": [
-              "image": "https://tile.loc.gov/image-services/iiif/service:amed:amedpllc:00406537456:0055/full/pct:6.25/0/default.jpg"
-            ]
-          }
-        ]
-      }
-    }
-
-    You would end up with a CSV with three columns (id, title and thumbnail) and two rows.
-    """
-
-    container = "dataframe"
-    name = "custom_json"
-    version = "0.0.1"
-    partition_access = True
-
-    def __init__(self, collection_url, metadata={}, csv_kwargs={}):
+import pandas as pd
+from typing import Any, Optional, Generator
+
+
+container = "dataframe"
+name = "custom_json"
+version = "0.0.1"
+partition_access = True
+
+
+class JsonSource(intake.source.base.DataSource):
+    def __init__(
+        self,
+        collection_url,
+        nextpage_path=None,
+        increment=None,
+        result_count=None,
+        record_selector=None,
+        dtype=None,
+        metadata=None,
+        wait=None,
+    ):
         super(JsonSource, self).__init__(metadata=metadata)
         self.collection_url = collection_url
-        self.metadata = metadata
-        self.csv_kwargs = csv_kwargs
-        self._setup_json_paths()
-
-    def read(self) -> DataFrame:
-        self._load_metadata()
-        df = self._get_dataframe()
-        return df
-
-    def _get_schema(self) -> Schema:
-        return Schema(
+        self.nextpage_path = nextpage_path
+        self.increment = increment
+        self.result_count = result_count
+        self.record_selector = record_selector
+        self.dtype = dtype
+        self.wait = wait
+        self._page_urls = []
+        self._path_expressions = {}
+        self.record_count = 0
+        self.record_limit = self.metadata.get("record_limit")
+
+    def _open_collection(self):
+        # either the API passes the next page or we increment with a url pattern
+        if self.nextpage_path:
```

Comment:
Am I reading the code properly here that we're building the list of `next` URLs which are then used by intake as the partition URLs?

Did you try simpifying this by using the `page_list` array instead of querying the API for each `url` and getting it's `next` value? That may help to reduce this complexity.